### PR TITLE
Implement CreateFeedback, UploadArtifact and CommitFeedback endpoints

### PR DESCRIPTION
**Pull request by tagboola**
Hey there! So you want to contribute to a Firebase SDK?
Before you file this pull request, please read these guidelines:

### Discussion

  * Read the contribution guidelines (CONTRIBUTING.md).
  * If this has been discussed in an issue, make sure to link to the issue here.
    If not, go file an issue about this **before creating a pull request** to discuss.

### Testing

  * Make sure all existing tests in the repository pass after your change.
  * If you fixed a bug or added a feature, add a new test to cover your code.

### API Changes

  * At this time we cannot accept changes that affect the public API.  If you'd like to help
    us make Firebase APIs better, please propose your change in a feature request so that we
    can discuss it together.

### CODE REVIEW

File: FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift [hunk 1]
```diff
@@ -63,12 +83,22 @@ struct FindReleaseResponse: Codable {
   var release: String
 }
 
+struct FeedbackReport: Codable {
+  var name: String?
+  var text: String
+}
+
+struct CreateFeedbackReportRequest: Codable {
```
[the remark this answers is not in the document]
Comment (tagboola) [reply FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift:91]: I think this is correctly named-- `CreateFeedbackReportRequest` is just the request message for making calls to the `CreateFeedbackReport`.

File: FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift [hunk 2]
```diff
@@ -200,6 +231,107 @@ struct FindReleaseResponse: Codable {
...
+          DispatchQueue.main.async {
+            completion(fadError)
+          }
+        }
+
+      uploadImageTask.resume()
+    }
+  }
+
+  public static func commitFeedback(feedbackName: String,
```
[the remark this answers is not in the document]
Comment (tagboola) [reply FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift]: Done.

File: FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift [hunk 3]
```diff
@@ -200,6 +231,107 @@ struct FindReleaseResponse: Codable {
...
+          DispatchQueue.main.async {
+            completion(feedback?.name, fadError)
+          }
+        }
+
+      createFeedbackTask.resume()
+    }
+  }
+
+  public static func uploadImage(feedbackName: String, image: UIImage,
```
[the remark this answers is not in the document]
Comment (tagboola) [reply FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift]: Done.

File: FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift [hunk 4]
```diff
@@ -200,6 +231,107 @@ struct FindReleaseResponse: Codable {
     }
   }
 
+  public static func createFeedback(releaseName: String, feedbackText: String,
```
[the remark this answers is not in the document]
Comment (tagboola) [reply FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift]: Done.

File: FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift [hunk 5]
```diff
@@ -200,6 +231,107 @@ struct FindReleaseResponse: Codable {
...
+                                    completion: @escaping AppDistributionCreateFeedbackCompletion) {
+    generateAuthToken { identifier, authTokenResult, error in
+      let urlString = String(
+        format: Strings.createFeedbackEndpointUrlTemplate,
+        releaseName
+      )
+      let request = createHttpRequest(
+        method: Strings.httpPost,
+        url: urlString,
+        authTokenResult: authTokenResult!
```
[the remark this answers is not in the document]
Comment (tagboola) [reply FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift]: Done.

File: FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift [hunk 6]
```diff
@@ -23,6 +23,13 @@ public typealias AppDistributionFetchReleasesCompletion = (_ releases: [Any]?, _
   -> Void
 public typealias AppDistributionFindReleaseCompletion = (_ releaseName: String?, _ error: Error?)
   -> Void
+public typealias AppDistributionCreateFeedbackCompletion = (_ feedbackName: String?,
```
Comment (paulb777) [inline FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift:26]: What's the reason for typealias's? We've usually found that, unlike typedefs in Objective-C, typealias's make code less readable in Swift, especially for callbacks where Xcode's code completion works less well.

Comment (tagboola) [reply FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift]: No resason-- this was just a relic of porting the existing code over from Objective-C. Inlined these and removed the typaliases.

**Review by tejasd — APPROVED** [review]

File: FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift [hunk 7]
```diff
@@ -159,6 +181,254 @@ struct FindReleaseResponse: Codable {
...
+                          urlSession: URLSession, displayVersion: String,
+                          buildVersion: String, codeHash: String,
+                          completion: @escaping (_ releaseName: String?, _ error: Error?)
+                            -> Void) {
+    generateAuthToken(installations: installations) { identifier, authTokenResult, error in
+      // TODO(tundeagboola) The backend may not accept project ID here in which case
+      // we'll have to figure out a way to get the project number
+      let urlString = String(
+        format: Strings.findReleaseEndpointUrlTemplate,
+        FirebaseApp.app()!.options.projectID!,
```
[the remark this answers is not in the document]
Comment (tagboola) [reply FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift]: Done.

File: FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift [hunk 8]
```diff
@@ -232,6 +502,24 @@ struct FindReleaseResponse: Codable {
...
+    guard let data = data else {
+      return nil
+    }
+    guard let mappedResponse = try? JSONDecoder().decode(T.self, from: data) else {
+      return nil
+    }
+    return mappedResponse
+  }
+
   static func handleReleaseResponse(data: NSData?, response: URLResponse?,
```
[the remark this answers is not in the document]
Comment (tagboola) [reply FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift:529]: For sure! I'm going to attempt to do this in a separate CL, so just adding a TODO for now.

File: FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift [hunk 9]
```diff
@@ -159,6 +181,254 @@ struct FindReleaseResponse: Codable {
...
+    generateAuthToken(installations: installations) { identifier, authTokenResult, error in
+      // TODO(tundeagboola) The backend may not accept project ID here in which case
+      // we'll have to figure out a way to get the project number
+      let urlString = String(
+        format: Strings.findReleaseEndpointUrlTemplate,
+        FirebaseApp.app()!.options.projectID!,
+        identifier!
+      )
+      guard var urlComponents = URLComponents(string: urlString) else {
+        // TODO(tundeagboola) Should I throw an exception here?
```
[the remark this answers is not in the document]
Comment (tagboola) [reply FirebaseAppDistributionInternal/Sources/AppDistributionApiService.swift]: Added logging and more specific TODO's around preferring exceptions versus piping errors.

**Pull request merged**